lib/ecto/parameterized_type.ex

```diff
@@ -0,0 +1,83 @@
+defmodule Ecto.ParameterizedType do
+  @moduledoc """
+  Parameterized types are similar to Ecto Custom Types, however they allow a set of options to be
+  specified in the schema which are initialized on compilation and passed to the callback functions
+  as a final argument. Note that options should be specified as a keyword, but it is idiomatic to
+  convert them to maps inside init for easier pattern matching in other callbacks.
+
+  ## Examples
+
+    To create a parameterized type, create a module as shown below:
+
+    defmodule MyApp.MyType do
+      use Ecto.ParameterizedType
+
+      def type(_opts), do: :string
+
+      def init(opts) do
+        ...
+        Enum.into(opts, %{})
+      end
+
+      def cast(data, current, opts) do
+        ...
+        cast_data
+      end
+
+      def load(data, opts) do
+        ...
+        {:ok, loaded_data}
+      end
+
+      def dump(data, opts) do
+        ...
+        {:ok, dumped_data}
+      end
+
+      def equal?(a, b, _opts) do
+        a == b
+      end
+    end
+
+  To use this type in a schema field, specify it in the form {module, keyword_opts} like this:
+
+    schema "foo" do
+      field "bar", {MyApp.MyType, opt1: :baz, opt2: :boo}
+    end
+
+  """
+
+  @type opts :: keyword() | map()
+
+  @callback init(opts :: keyword()) :: opts()
```

Comment:
I believe the field name would also be given on init, correct?

Reply:
Oh, I see you are passing it as an option. I am thinking it may be better to make this become: `init(schema, field, value)` and we don't put any restriction on the value. Thoughts?

Reply:
Sorry, maybe I do not see something, but what is the purpose of passing `schema` and `field` to any callback? I don't see a use case and therefore I don't believe anybody would need those. Also anything dynamic like `enum` values should be passed to `attrs`/`config`/`opts`/or however we name it, right?

Reply:
@Eiji7 enum definitely needs them, but we get to cheat because we have a macro for passing them in. So I can see others needing it in the future. The whole point of `init` is to receive all information so you decide what to keep. What you keep is what becomes `config`.

Reply:
Btw, if we decide to go with `field :foo, Ecto.Enum, values: [...]`, then I think this signature is perfect, as it will always be a keyword list. We can also pass the `:schema` and `:field` as an option too, the same as today. So I would just keep it as is. You can consider this particular comment resolved. :)